### PR TITLE
List API

### PR DESCRIPTION
**Pull request by JoshVanL**
Adds a new `List` API to the cron service. This API returns the list of current committed active jobs that have a name matching the given prefix. An empty prefix will return all jobs.

```go
List(ctx context.Context, prefix string) (*ListResponse, error)
```

A new `ListResponse` proto message has been added which contains a list of named Jobs.

As part of this change to clean up the codebase, the cron API definition has been moved into `api/api.go`. A new `queue` runnable has been created which is responsible for managing the underlying job queue based on events from the informer. A new `internal/api` package is responsible for implementing the server CRUD APIs.

Uses dapr/kit concurrency/map from https://github.com/dapr/kit/pull/104

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -78,3 +78,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/dapr/kit => github.com/joshvanl/kit v0.0.0-20240923214824-8b9ec753c4bf
```
Comment (cicoyle) [inline go.mod:82]: this got merged, so we can use a commit from kit now

File: internal/api/validator/validator.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+// Validator validates API request payloads.
+type Validator struct {
+	jobNameSanitizer *strings.Replacer
+}
+
+func New(opts Options) *Validator {
+	jobNameSanitizer := opts.JobNameSanitizer
+	if jobNameSanitizer == nil {
+		jobNameSanitizer = strings.NewReplacer("_", "", ":", "", "-", "")
```
Comment (cicoyle) [inline internal/api/validator/validator.go:31]:
```suggestion
		jobNameSanitizer = strings.NewReplacer("_", "", ":", "", "-", "", " ", "")
```
if we add this I think we fix a discord user issue where they saw an err due to the job name (reminder/actor I believe) having a space in it.

File: proto/list.proto [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+// ListResponse is returned by the List RPC.
+message ListResponse {
+  // jobs is the list of jobs.
+  repeated NamedJob jobs = 1;
+
+  // more indicates there are more jobs in the given prefix which do not fit in
+  // the response size.
+	bool more = 2;
```
Comment (cicoyle) [inline proto/list.proto:29]: nit spacing

Comment (cicoyle) [reply proto/list.proto:29]: this handles pagination?

File: api/api.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	// Delete deletes a job from the cron instance.
+	Delete(ctx context.Context, name string) error
+
+	// DeletePrefixes deletes all jobs with the given prefixes from the cron
+	// instance.
+	DeletePrefixes(ctx context.Context, prefixes ...string) error
+
+	// List lists all jobs under a given job name prefix.
+	List(ctx context.Context, prefix string) (*ListResponse, error)
```
Comment (cicoyle) [inline api/api.go:35]: Changes mostly lgtm, but its a bit hard to parse for the changes for the list api given the additional refactoring/nix changes. Is there a test for the api in this PR?

File: go.mod (hunk 1 above)
Comment (JoshVanL) [reply go.mod]: Thanks :+1: updated

File: internal/api/validator/validator.go (hunk 2 above)
Comment (JoshVanL) [reply internal/api/validator/validator.go:31]: Thanks, makes sense to me- instead of adding it here (to an already bloated PR), I've made [this issue](https://github.com/diagridio/go-etcd-cron/issues/46) so we can track it separately and be intentional about the test cases we are adding with a new PR. There is also the open question about our current key segment concatenation and breaking changes https://github.com/diagridio/go-etcd-cron/pull/45

File: proto/list.proto (hunk 3 above)
Comment (JoshVanL) [reply proto/list.proto]: Thanks :+1: 

No, we don't handle any kind of pagination or filtering etc. We simply request the most amount we can from Etcd. I updated the [API call](https://github.com/diagridio/go-etcd-cron/pull/43/files#diff-5b9b550fff634f1fd5596893d63ceeb6afa6a355427db919a644a21f097b7834R190) to ensure we get the upper limit using a [parameter of `0`](https://pkg.go.dev/go.etcd.io/etcd/client/v3#WithLimit). Indeed, the client is out of luck if their request is larger than the limit for now. We can add further parameters on the `List` request to resolve this issue in future, though I thought that at least adding `more`, the client has a chance of understand what's occurring.

File: api/api.go (hunk 4 above)
Comment (JoshVanL) [reply api/api.go:35]: My bad, sorry about that- got semi carried away with it :grimacing: 

Here are the two places where we are testing the `List` API specifically. Happy to separate the refactor and API addition if that would make review easier.

https://github.com/diagridio/go-etcd-cron/pull/43/files#diff-4d514085b47744fcdfdda631f3b5e141857e34feb020911194a877879fcf1501R209
https://github.com/diagridio/go-etcd-cron/pull/43/files#diff-efc48b37700cd7674ce96c446ce0043c8f3d55888ddba4392b897f6411e2c30dR231

File: internal/api/validator/validator.go (hunk 2 above)
Comment (cicoyle) [reply internal/api/validator/validator.go:31]: sure thing.

File: internal/api/serve_test.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+		t.Parallel()
+
+		cron := cron.SinglePartitionRun(t)
+
+		resp, err := cron.List(context.Background(), "")
+		require.NoError(t, err)
+		assert.Empty(t, resp.Jobs)
+	})
+
+	t.Run("List should return jobs which are in the namespace", func(t *testing.T) {
```
Comment (cicoyle) [inline internal/api/serve_test.go:244]: This test is good for now. Have you done any manual validation as to what happens when the list returns too large of a list of jobs? I imagine we want to do a quick sanity check on this for the auto migration of reminders.

Comment (cicoyle) [reply internal/api/serve_test.go:244]: Other than this last comment, the PR lgtm

Comment (JoshVanL) [reply internal/api/serve_test.go:244]: Thanks for picking this up. I've been looking into the code and documentation and as far as I see, `more` is only relevant in the circumstances where a `limit` is set and the response list exceeds that. Since we set no limit (`limit=0`), this value will never be `true`. I'm going to go ahead and remove it from the protos. Thanks again!

**Review by cicoyle — APPROVED** [review]

**Review by artursouza — APPROVED** [review]
